File: java/parser/src/main/java/gov/lbl/parser/parser/Stored_definitionVisitor.java
package gov.lbl.parser.parser;

import java.util.List;

import org.antlr.v4.runtime.tree.ParseTree;
import org.antlr.v4.runtime.tree.TerminalNode;

import gov.lbl.antlr4.visitor.modelicaBaseVisitor;
import gov.lbl.antlr4.visitor.modelicaParser;
import gov.lbl.parser.domain.Class_definition;
import gov.lbl.parser.domain.Final_class_definition;
import gov.lbl.parser.domain.Name;

import java.util.ArrayList;

import gov.lbl.parser.domain.Stored_definition;

public class Stored_definitionVisitor extends modelicaBaseVisitor<Stored_definition> {
  @Override
  public Stored_definition visitStored_definition(modelicaParser.Stored_definitionContext ctx) {
    NameVisitor nameVisitor = new NameVisitor();
    Class_definitionVisitor class_definitionVisitor = new Class_definitionVisitor();

    Name name = ctx.name() == null ? null : ctx.name().accept(nameVisitor);
    String within = name.getName();
    List<Final_class_definition> final_class_definitions = new ArrayList<>();

    List<ParseTree> children = ctx.children;
    Boolean prev_is_final = false;

    for (ParseTree o : children) {            
        if ((o instanceof modelicaParser.Class_definitionContext)) {   
          Class_definition class_definition = o.accept(class_definitionVisitor);
          final_class_definitions.add(new Final_class_definition(prev_is_final, class_definition));
          prev_is_final = false;
        } else if ((o instanceof TerminalNode)) {
            if (o.getText().equals("final")) {
                prev_is_final = true;
            }
        }
    }
    
    return new Stored_definition(within, final_class_definitions);
  }
}

File: jsParser/parser/Stored_definitionVisitor.js
const modelicaVisitor = require('../antlrFiles/modelicaVisitor').modelicaVisitor
const Stored_definition = require('../domain/Stored_definition')
const Final_class_definition = require('../domain/Final_class_definition')
const Name = require('../domain/Name')
const Class_definition = require('../domain/Class_definition')

const NameVisitor = require('./NameVisitor')
const { modelicaParser } = require('../antlrFiles/modelicaParser')
const { TerminalNode } = require('antlr4/tree/Tree')
const Class_definitionVisitor = require('./Class_definitionVisitor')

const util = require('util')

class Stored_definitionVisitor {
  constructor () {
    modelicaVisitor.call(this)
    return this
  }

  visitStored_definition (ctx) {
    let within = ''
    const final_class_definitions = []

    if (ctx.name()) {
      const nameVisitor = new NameVisitor.NameVisitor()
      const name = nameVisitor.visitName(ctx.name())

      name.name_parts.forEach(name_part => {
        if (name_part.dot_op) {
          within = util.format('%s%s', within, '.')
        }

        if (name_part.identifier) {
          within = util.format('%s%s', within, name_part.identifier)
        }
      })
    }

    const class_definitionVisitor = new Class_definitionVisitor.Class_definitionVisitor()
    let prev_final = false
    ctx.children.forEach(child => {
      if (child instanceof modelicaParser.Class_definitionContext) {
        const class_definition = class_definitionVisitor.visitClass_definition(child)
        final_class_definitions.push(new Final_class_definition.Final_class_definition(prev_final, class_definition))
        prev_final = false
      } else if (child instanceof TerminalNode) {
        if (child.getText() == 'final') {
          prev_final = true
        }
      }
    })

    return new Stored_definition.Stored_definition(within, final_class_definitions)
  }
};

Stored_definitionVisitor.prototype = Object.create(modelicaVisitor.prototype)

exports.Stored_definitionVisitor = Stored_definitionVisitor
exports.visitStored_definition = this.visitStored_definition
